src/Database/LSMTree/Internal/RefCount.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE MagicHash #-}
+
+module Database.LSMTree.Internal.RefCount (
+    RefCounter
+  , unsafeMkRefCounterN
+  , mkRefCounterN
+  , mkRefCounter1
+  , RefCountError (..)
+  , addReference
+  , readRefCount
+  , removeReference
+  , replaceFinaliser
+  , finalise
+  , RefCount (..)
+  ) where
+
+import           Control.DeepSeq
+import           Control.Monad
+import           Control.Monad.Class.MonadThrow
+import           Control.Monad.Primitive
+import           Data.Maybe
+import           Data.Primitive.MutVar
+
+-- | A reference counter with an optional finaliser action. Once the reference
+-- count reaches @0@, the finaliser will be run.
+data RefCounter m = RefCounter {
+    countVar     :: !(MutVar (PrimState m) RefCount)
```

Comment:
This could be lower level: `PrimVar (PrimState m) Int`. Using the `Data.Primitive.PrimVar` module from the `primitive` package.

This would mean no allocation for mutations (otherwise we allocate `Int` boxes) and allows low level atomic operations like `casInt`, `fetchAddInt` and `fetchSubInt`.